### PR TITLE
Add address and full node string

### PR DESCRIPTION
**Pull request by lucusfly**
example node id: snode://0x9484798ab92ad49dddb961535bae1320c688a633f52bc8d9fadf667b5bba3cf4675ec066cbeca6cd2005540fdf0fff1184718a0a0125159088ee87dd4970b614@:9000

### CODE REVIEW

File: cmd/discovery/cmd/send.go [hunk 1]
```diff
@@ -7,17 +7,17 @@ package cmd
 
 import (
 	"fmt"
-
-	"github.com/seeleteam/go-seele/p2p/discovery"
+	"net"
 
 	"github.com/spf13/cobra"
+
```
Comment (WTer) [inline cmd/discovery/cmd/send.go]: remove this blank line.

File: cmd/discovery/cmd/listen.go [hunk 2]
```diff
@@ -25,7 +25,18 @@ to quickly create a Cobra application.`,
 	Run: func(cmd *cobra.Command, args []string) {
 		fmt.Println("listen called")
 
-		discovery.StartService("9000", *id)
+		// use a fixed id for test
+		id := "snode://c03ff3c956d0a320b153a097c3d04efa488d43d7d7e05a44791492c9979ff558f9956c0a6b0c414783476f02ad8557349d35ba9373dadfa9a7a44fd88328189f@:9000"
```
Comment (WTer) [inline cmd/discovery/cmd/listen.go]: why don't you pass this id form cmdline?

File: cmd/discovery/cmd/send.go [hunk 3]
```diff
@@ -7,17 +7,17 @@ package cmd
...
 )
 
-var
-(
-	myport *string
-	targetid *string
-	targetport *string
+var (
+	addr          *string
+	bootstrapNode *string
```
Comment (WTer) [inline cmd/discovery/cmd/send.go]: nice to add comments for this var and what the difference between addr and bootstrapNode

File: cmd/discovery/cmd/send.go [hunk 4]
```diff
@@ -32,18 +32,37 @@ This application is a tool to generate the needed files
 to quickly create a Cobra application.`,
 	Run: func(cmd *cobra.Command, args []string) {
 		fmt.Println("send called")
-		fmt.Println(*id)
 
-		discovery.SendPing(*myport, *targetid, *targetport)
+		nodeId, err := discovery.NewNodeFromString(*bootstrapNode)
+		if err != nil {
+			fmt.Println(err)
```
Comment (WTer) [inline cmd/discovery/cmd/send.go]: use log

Comment (WTer) [reply cmd/discovery/cmd/send.go]: this is cmdline tool? if yes, console output should be ok.

File: cmd/discovery/cmd/listen.go (hunk 2 above)
Comment (lucusfly) [reply cmd/discovery/cmd/listen.go]: fix id for test, so I could use the same command line

File: cmd/discovery/cmd/send.go [hunk 5]
```diff
@@ -32,18 +32,37 @@ This application is a tool to generate the needed files
...
+			return
+		}
+
+		myAddr, err := net.ResolveUDPAddr("udp", *addr)
```
Comment (WTer) [inline cmd/discovery/cmd/send.go]: "udp" is case-sensitive?
and better to define it as constant.

File: cmd/discovery/cmd/listen.go (hunk 2 above)
Comment (lucusfly) [reply cmd/discovery/cmd/listen.go]: and it's for test, so I don't want pass it by cmd line

File: cmd/discovery/cmd/send.go [hunk 6]
```diff
@@ -32,18 +32,37 @@ This application is a tool to generate the needed files
...
+			return
+		}
+
+		myId, err := common.GenerateRandomAddress()
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+
+		mynode := discovery.NewNodeWithAddr(*myId, myAddr)
```
Comment (WTer) [inline cmd/discovery/cmd/send.go]: so I suppose you try to bind the physical address myAddr with myId that is node identifier?

File: cmd/discovery/cmd/send.go [hunk 7]
```diff
@@ -32,18 +32,37 @@ This application is a tool to generate the needed files
...
 	},
 }
 
 func init() {
 	rootCmd.AddCommand(sendCmd)
 
-	myport = sendCmd.Flags().String("myport", "9001", "my port")
-	targetid = sendCmd.Flags().String("targetid", "", "target id")
-	targetport = sendCmd.Flags().String("targetport", "9000", "target port")
+	addr = sendCmd.Flags().StringP("addr", "a", ":9001", "node addr")
```
Comment (WTer) [inline cmd/discovery/cmd/send.go]: Flags is a little confused from its name, can we rename it?

File: common/Address.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+// Bytes get the actual bytes
+func (id *Address) Bytes() []byte {
+	return id[:]
+}
+
+// ToSha get the node hash
+func (id *Address) ToSha() *Hash {
+	data := crypto.Keccak256Hash(id[:])
+
```
Comment (WTer) [inline common/Address.go]: this blank line can be removed.

File: cmd/discovery/cmd/send.go (hunk 5 above)
Comment (lucusfly) [reply cmd/discovery/cmd/send.go]: this is a golang buildin method, I don't think we need to move it to constant

File: cmd/discovery/cmd/send.go (hunk 6 above)
Comment (lucusfly) [reply cmd/discovery/cmd/send.go]: yes

File: cmd/discovery/cmd/send.go (hunk 7 above)
Comment (lucusfly) [reply cmd/discovery/cmd/send.go]: flags is what golang cmd line use

File: common/Address.go (hunk 8 above)
Comment (lucusfly) [reply common/Address.go]: remove

File: cmd/discovery/cmd/send.go (hunk 1 above)
Comment (lucusfly) [reply cmd/discovery/cmd/send.go]: remove

File: crypto/sha.go [hunk 9]
```diff
@@ -5,18 +5,18 @@
...
 // Keccak256Hash calculates and returns the Keccak256 hash of the input data,
 // converting it to an internal Hash data structure.
-func Keccak256Hash(data ...[]byte) (h common.Hash) {
+func Keccak256Hash(data ...[]byte) []byte {
 	d := sha3.NewKeccak256()
 	for _, b := range data {
 		d.Write(b)
 	}
 
+	h := make([]byte, 32)
```
Comment (WTer) [inline crypto/sha.go]: nice to define this magic number to a constant.

File: crypto/sha.go [hunk 10]
```diff
@@ -5,18 +5,18 @@
...
 	d.Sum(h[:0])
 	return h
-}
+}
```
Comment (WTer) [inline crypto/sha.go]: take care

File: crypto/sha.go [hunk 11]
```diff
@@ -5,18 +5,18 @@
...
 // converting it to an internal Hash data structure.
-func Keccak256Hash(data ...[]byte) (h common.Hash) {
+func Keccak256Hash(data ...[]byte) []byte {
 	d := sha3.NewKeccak256()
 	for _, b := range data {
 		d.Write(b)
 	}
 
+	h := make([]byte, 32)
 	d.Sum(h[:0])
```
Comment (WTer) [inline crypto/sha.go]: h[: 0]? seems should be h[:]

File: p2p/discovery/bucket_test.go [hunk 12]
```diff
@@ -40,9 +41,13 @@ func Test_Bucket(t *testing.T) {
 }
 
 func getNode(port string) *Node {
-	id := getRandomNodeID()
+	id, err := common.GenerateRandomAddress()
+	if err != nil {
+		panic(err)
+	}
+
 	addr, _ := net.ResolveUDPAddr("udp", ":"+port)
```
Comment (WTer) [inline p2p/discovery/bucket_test.go]: nice to extract a common to represent this addr combination logic:
"udp", ":"+port

File: p2p/discovery/node_test.go [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+import (
+	"encoding/hex"
+	"testing"
+
+	"github.com/magiconair/properties/assert"
+)
+
+func Test_NodeId(t *testing.T) {
+	id := "snode://c03ff3c956d0a320b153a097c3d04efa488d43d7d7e05a44791492c9979ff558f9956c0a6b0c414783476f02ad8557349d35ba9373dadfa9a7a44fd88328189f@192.168.122.132:9000"
```
Comment (WTer) [inline p2p/discovery/node_test.go]: split id to publickey, addr and port,
so the following assert compared be more convenient

File: crypto/sha.go (hunk 9 above)
Comment (lucusfly) [reply crypto/sha.go]: good suggestion

File: crypto/sha.go (hunk 10 above)
Comment (lucusfly) [reply crypto/sha.go]: done

File: crypto/sha.go (hunk 11 above)
Comment (lucusfly) [reply crypto/sha.go]: no, sum will change its parameter

File: cmd/discovery/cmd/listen.go (hunk 2 above)
Comment (WTer) [reply cmd/discovery/cmd/listen.go]: nice.

File: p2p/discovery/bucket_test.go (hunk 12 above)
Comment (lucusfly) [reply p2p/discovery/bucket_test.go]: this will be temp, we will add ip part later

File: p2p/discovery/node_test.go (hunk 13 above)
Comment (lucusfly) [reply p2p/discovery/node_test.go]: it is done in NewNodeFromString()

File: cmd/discovery/cmd/send.go (hunk 4 above)
Comment (lucusfly) [reply cmd/discovery/cmd/send.go]: yes

**Pull request merged**